### ISSUE TITLE
questions and help about libtorch under multi thread

### ISSUE THREAD

**Issue by wanggujin** [pytorch/pytorch#23920]
i use libtorch v1.1.0 in c++,
but in multi thread program,  the program will core dump in this funcion:
```cpp
void ModelInference(const std::vector<torch::jit::IValue>& torch_inputs,
    std::shared_ptr<torch::jit::script::Module> torch_model_ptr,
    LaneSequence* lane_sequence_ptr) {
  auto torch_output_tuple = torch_model_ptr->forward(torch_inputs).toTuple();
   auto probability_tensor = torch_output_tuple->elements()[0].toTensor();
   auto finish_time_tensor = torch_output_tuple->elements()[1].toTensor();
}
```
some error message as follows:
```
terminate called after throwing an instance of 'c10::Error'
  what():  method 'forward' already defined. (insert at libtorch/include/torch/csrc/jit/script/module.h:515)
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0x6a (0x7ffff49335ca in libtorch/lib/libc10.so)
frame #1: std::unique_ptr<torch::jit::script::Method, std::default_delete<torch::jit::script::Method> >& torch::jit::script::Module::insert<std::unique_ptr<torch::jit::script::Method, std::default_delete<torch::jit::script::Method> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::unique_ptr<torch::jit::script::Method, std::default_delete<torch::jit::script::Method> >, std::allocator<std::unique_ptr<torch::jit::script::Method, std::default_delete<torch::jit::script::Method> > > >&, torch::jit::script::Module::EntityType, std::unique_ptr<torch::jit::script::Method, std::default_delete<torch::jit::script::Method> >) + 0x20a 
frame #2: torch::jit::script::Module::find_method(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const + 0x167 
frame #3: torch::jit::script::Module::get_method(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const + 0x35 
frame #4: torch::jit::script::Module::forward(std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0xb3 
frame #5: CruiseMLPEvaluator::ModelInference(std::vector<c10::IValue, std::allocator<c10::IValue> > const&, std::shared_ptr<torch::jit::script::Module>, LaneSequence*) + 0x5d
```
someone knows that's why? how can i solve this problem?

**Comment by ailzhang** [comment]
can you give a repro script? Thanks!

**Comment by dmcconachie** [comment]
As far as I can tell, `Module::forward` is not threadsafe, probably for multiple reasons. Some could be `trivial` such as simple lack of const declarations https://github.com/pytorch/pytorch/issues/23901, others a more core challenge due to the way torch works. I.e.; does `forward(...)` cache computation that is needed to do backprop? Probably (I don't actually know).

**Comment by dzhulgakov** [comment]
It's indeed a race condition in getting the method. It was fixed in #21398 and will be out in 1.2 (in a week or so). As a workaround for now - you can call get_method() once non-concurrently as a part of initialization and it should fix the issue.

In general, TorchScript execution is threadsafe (unless your TorchScript program does something non-thread-safe like changing the same parameter, but that's rare)

**Comment by eellison** [comment]
According to @dzhulgakov this was closed with https://github.com/pytorch/pytorch/pull/21398, if you are still experiencing this please re-open.
